org.eclipse.jdt.core/codeassist/org/eclipse/jdt/internal/codeassist/complete/CompletionParser.java

```diff
@@ -4176,6 +4181,12 @@ protected void consumeToken(int token) {
 					popElement(K_BETWEEN_LEFT_AND_RIGHT_BRACKET);
 				}
 				break;
+			case TokenNameARROW, TokenNameCOLON:
+				if(topKnownElementKind(COMPLETION_OR_ASSIST_PARSER) == K_RECORD_PATTERN) {
```

Comment:
Running `CompletionTestsForRecordPattern` with a breakpoint here, I can see various situations where `K_RECORD_PATTERN` is on the stack, but it's not found because it's not the top element. I guess this produces stale stack state, which might cause trouble later.

Top elements are `K_SELECTOR_QUALIFIER`, `K_SELECTOR_INVOCATION_TYPE` and `K_SELECTOR`. Since these tokens don't make sense here (we don't have an invocation), perhaps we should suppress those tokens when `K_RECORD_PATTERN` is on top?

Reply:
I tried, then the test fails, here is the parser output

```
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (class)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) ClassHeaderName1
ClassHeaderName1 ::= Modifiersopt class Identifier
----------------------------------------
Reduce       - ClassHeaderName
----------------------------------------
Reduce       - ClassHeaderExtendsopt
----------------------------------------
Reduce       - ClassHeaderImplementsopt
----------------------------------------
Reduce       - ClassHeaderPermittedSubclassesopt
             - ClassHeader
ClassHeader ::= ClassHeaderName ClassHeaderExtendsopt...
----------------------------------------
Shift        - (public)
Reduce       - NestedType
NestedType ::=
----------------------------------------
Shift        - (void)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - (()
Shift/Reduce - (Identifier) MethodHeaderName
MethodHeaderName ::= Modifiersopt Type Identifier LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (Identifier) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - VariableDeclaratorIdOrThis
VariableDeclaratorIdOrThis ::= VariableDeclaratorId
             - FormalParameter
FormalParameter ::= Modifiersopt Type...
----------------------------------------
Reduce       - FormalParameterListopt
----------------------------------------
Shift/Reduce - ({) MethodHeaderRightParen
MethodHeaderRightParen ::= RPAREN
----------------------------------------
Reduce       - Dimsopt
Dimsopt ::=
             - MethodHeaderExtendedDims
MethodHeaderExtendedDims ::= Dimsopt
----------------------------------------
Reduce       - MethodHeaderThrowsClauseopt
             - MethodHeader
MethodHeader ::= MethodHeaderName FormalParameterListopt
----------------------------------------
Reduce       - NestedMethod
NestedMethod ::=
----------------------------------------
Shift        - (})
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
----------------------------------------
Shift/Reduce - (}) MethodBody
MethodBody ::= NestedMethod LBRACE BlockStatementsopt...
             - MethodDeclaration
MethodDeclaration ::= MethodHeader MethodBody
----------------------------------------
Reduce       - ClassBodyDeclarations
----------------------------------------
Reduce       - ClassBodyDeclarationsopt
ClassBodyDeclarationsopt ::= NestedType...
----------------------------------------
Shift/Reduce - (Unexpected End Of File) ClassBody
             - ClassDeclaration
ClassDeclaration ::= ClassHeader ClassBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (switch)
Shift        - (()
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
----------------------------------------
Shift        - ({)
Reduce       - PostExpressionInSwitchStatement
PostExpressionInSwitchStatement ::=
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (case)
Shift        - (BeginCaseElement)
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (()
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift/Reduce - (Identifier) PushLPAREN
PushLPAREN ::= LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - (,) TypePattern
TypePattern ::= Modifiersopt Type Identifier
----------------------------------------
Reduce       - ComponentPatternList
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - ()) TypePattern
TypePattern ::= Modifiersopt Type Identifier
             - ComponentPatternList
ComponentPatternList ::= ComponentPatternList COMMA...
----------------------------------------
Reduce       - ComponentPatternListopt
----------------------------------------
Shift/Reduce - (BeginCaseExpr) PushRPAREN
PushRPAREN ::= RPAREN
             - RecordPattern
RecordPattern ::= Modifiersopt ReferenceType PushLPAREN
             - CaseLabelElementPattern
----------------------------------------
Reduce       - CaseLabelElement
CaseLabelElement ::= CaseLabelElementPattern
----------------------------------------
Reduce       - SwitchLabelCaseLhs
SwitchLabelCaseLhs ::= case CaseLabelElements
----------------------------------------
Shift        - (->)
Shift/Reduce - ({) SwitchLabelExpr
SwitchLabelExpr ::= SwitchLabelCaseLhs BeginCaseExpr...
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (Identifier)
Shift        - (})
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
Shift        - (})
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (Unexpected End Of File)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (RestrictedIdentifierrecord)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (() RecordHeaderName1
RecordHeaderName1 ::= Modifiersopt...
----------------------------------------
Reduce       - RecordHeaderName
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - (,)
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - RecordComponent
RecordComponent ::= Modifiersopt Type...
----------------------------------------
Reduce       - RecordComponents
----------------------------------------
Shift        - (int)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - RecordComponent
RecordComponent ::= Modifiersopt Type...
             - RecordComponents
RecordComponents ::= RecordComponents COMMA...
----------------------------------------
Reduce       - RecordComponentsopt
----------------------------------------
Shift/Reduce - (implements) RecordComponentHeaderRightParen
RecordComponentHeaderRightParen ::= RPAREN
             - RecordHeader
RecordHeader ::= LPAREN RecordComponentsopt...
----------------------------------------
Shift        - (Identifier)
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - InterfaceType
InterfaceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - ClassHeaderImplements
ClassHeaderImplements ::= implements InterfaceTypeList
             - RecordHeaderPart
RecordHeaderPart ::= RecordHeaderName RecordHeader...
----------------------------------------
Shift        - (})
Reduce       - RecordBodyDeclarationopt
RecordBodyDeclarationopt ::=
----------------------------------------
Shift/Reduce - (Unexpected End Of File) RecordBody
RecordBody ::= LBRACE RecordBodyDeclarationopt RBRACE
             - RecordDeclaration
RecordDeclaration ::= RecordHeaderPart RecordBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (class)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) ClassHeaderName1
ClassHeaderName1 ::= Modifiersopt class Identifier
----------------------------------------
Reduce       - ClassHeaderName
----------------------------------------
Reduce       - ClassHeaderExtendsopt
----------------------------------------
Reduce       - ClassHeaderImplementsopt
----------------------------------------
Reduce       - ClassHeaderPermittedSubclassesopt
             - ClassHeader
ClassHeader ::= ClassHeaderName ClassHeaderExtendsopt...
----------------------------------------
Shift        - (public)
Reduce       - NestedType
NestedType ::=
----------------------------------------
Shift        - (void)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - (()
Shift/Reduce - (Identifier) MethodHeaderName
MethodHeaderName ::= Modifiersopt Type Identifier LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (Identifier) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - VariableDeclaratorIdOrThis
VariableDeclaratorIdOrThis ::= VariableDeclaratorId
             - FormalParameter
FormalParameter ::= Modifiersopt Type...
----------------------------------------
Reduce       - FormalParameterListopt
----------------------------------------
Shift/Reduce - ({) MethodHeaderRightParen
MethodHeaderRightParen ::= RPAREN
----------------------------------------
Reduce       - Dimsopt
Dimsopt ::=
             - MethodHeaderExtendedDims
MethodHeaderExtendedDims ::= Dimsopt
----------------------------------------
Reduce       - MethodHeaderThrowsClauseopt
             - MethodHeader
MethodHeader ::= MethodHeaderName FormalParameterListopt
----------------------------------------
Reduce       - NestedMethod
NestedMethod ::=
----------------------------------------
Shift        - (})
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
----------------------------------------
Shift/Reduce - (}) MethodBody
MethodBody ::= NestedMethod LBRACE BlockStatementsopt...
             - MethodDeclaration
MethodDeclaration ::= MethodHeader MethodBody
----------------------------------------
Reduce       - ClassBodyDeclarations
----------------------------------------
Reduce       - ClassBodyDeclarationsopt
ClassBodyDeclarationsopt ::= NestedType...
----------------------------------------
Shift/Reduce - (Unexpected End Of File) ClassBody
             - ClassDeclaration
ClassDeclaration ::= ClassHeader ClassBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (switch)
Shift        - (()
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
----------------------------------------
Shift        - ({)
Reduce       - PostExpressionInSwitchStatement
PostExpressionInSwitchStatement ::=
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (case)
Shift        - (BeginCaseElement)
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (()
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift/Reduce - (Identifier) PushLPAREN
PushLPAREN ::= LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - (,) TypePattern
TypePattern ::= Modifiersopt Type Identifier
----------------------------------------
Reduce       - ComponentPatternList
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - ()) TypePattern
TypePattern ::= Modifiersopt Type Identifier
             - ComponentPatternList
ComponentPatternList ::= ComponentPatternList COMMA...
----------------------------------------
Reduce       - ComponentPatternListopt
----------------------------------------
Shift/Reduce - (BeginCaseExpr) PushRPAREN
PushRPAREN ::= RPAREN
             - RecordPattern
RecordPattern ::= Modifiersopt ReferenceType PushLPAREN
             - CaseLabelElementPattern
----------------------------------------
Reduce       - CaseLabelElement
CaseLabelElement ::= CaseLabelElementPattern
----------------------------------------
Reduce       - SwitchLabelCaseLhs
SwitchLabelCaseLhs ::= case CaseLabelElements
----------------------------------------
Shift        - (->)
Shift/Reduce - ({) SwitchLabelExpr
SwitchLabelExpr ::= SwitchLabelCaseLhs BeginCaseExpr...
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (Identifier)
Shift        - (})
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Restart      -
Shift        - (switch)
Shift        - (()
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
----------------------------------------
Shift        - ({)
Reduce       - PostExpressionInSwitchStatement
PostExpressionInSwitchStatement ::=
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (case)
Restart      -
Shift        - (BeginCaseElement)
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (Identifier)
Shift        - (()
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
             - ArgumentList
----------------------------------------
Reduce       - ArgumentListopt
----------------------------------------
Error        -
Shift        - (()
Shift/Reduce - (Identifier) PushLPAREN
PushLPAREN ::= LPAREN
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
Shift        - (Identifier)
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - PushModifiers
PushModifiers ::=
----------------------------------------
Shift        - (,)
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
----------------------------------------
Reduce       - EnterVariable
EnterVariable ::=
----------------------------------------
Reduce       - ExitVariableWithoutInitialization
ExitVariableWithoutInitialization ::=
             - VariableDeclarator
----------------------------------------
Shift        - (Identifier)
Shift        - (Identifier)
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
----------------------------------------
Reduce       - EnterVariable
EnterVariable ::=
----------------------------------------
Reduce       - ExitVariableWithoutInitialization
ExitVariableWithoutInitialization ::=
             - VariableDeclarator
             - VariableDeclarators
VariableDeclarators ::= VariableDeclarators COMMA...
----------------------------------------
Reduce       - LocalVariableDeclaration
LocalVariableDeclaration ::= Type PushModifiers...
----------------------------------------
Error        -
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
Shift        - ())
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (->)
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - ({)
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (Identifier)
Restart      -
Shift        - (Identifier)
Shift        - (Unexpected End Of File)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Restart      -
Shift        - (})
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (})
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (Unexpected End Of File)
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
name[LOCAL_VARIABLE_REF]{name, null, Ljava.lang.Object;, name, null, 52}
```

Reply:
This is when not excluded

```
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (class)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) ClassHeaderName1
ClassHeaderName1 ::= Modifiersopt class Identifier
----------------------------------------
Reduce       - ClassHeaderName
----------------------------------------
Reduce       - ClassHeaderExtendsopt
----------------------------------------
Reduce       - ClassHeaderImplementsopt
----------------------------------------
Reduce       - ClassHeaderPermittedSubclassesopt
             - ClassHeader
ClassHeader ::= ClassHeaderName ClassHeaderExtendsopt...
----------------------------------------
Shift        - (public)
Reduce       - NestedType
NestedType ::=
----------------------------------------
Shift        - (void)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - (()
Shift/Reduce - (Identifier) MethodHeaderName
MethodHeaderName ::= Modifiersopt Type Identifier LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (Identifier) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - VariableDeclaratorIdOrThis
VariableDeclaratorIdOrThis ::= VariableDeclaratorId
             - FormalParameter
FormalParameter ::= Modifiersopt Type...
----------------------------------------
Reduce       - FormalParameterListopt
----------------------------------------
Shift/Reduce - ({) MethodHeaderRightParen
MethodHeaderRightParen ::= RPAREN
----------------------------------------
Reduce       - Dimsopt
Dimsopt ::=
             - MethodHeaderExtendedDims
MethodHeaderExtendedDims ::= Dimsopt
----------------------------------------
Reduce       - MethodHeaderThrowsClauseopt
             - MethodHeader
MethodHeader ::= MethodHeaderName FormalParameterListopt
----------------------------------------
Reduce       - NestedMethod
NestedMethod ::=
----------------------------------------
Shift        - (})
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
----------------------------------------
Shift/Reduce - (}) MethodBody
MethodBody ::= NestedMethod LBRACE BlockStatementsopt...
             - MethodDeclaration
MethodDeclaration ::= MethodHeader MethodBody
----------------------------------------
Reduce       - ClassBodyDeclarations
----------------------------------------
Reduce       - ClassBodyDeclarationsopt
ClassBodyDeclarationsopt ::= NestedType...
----------------------------------------
Shift/Reduce - (Unexpected End Of File) ClassBody
             - ClassDeclaration
ClassDeclaration ::= ClassHeader ClassBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (switch)
Shift        - (()
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
----------------------------------------
Shift        - ({)
Reduce       - PostExpressionInSwitchStatement
PostExpressionInSwitchStatement ::=
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (case)
Shift        - (BeginCaseElement)
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (()
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift/Reduce - (Identifier) PushLPAREN
PushLPAREN ::= LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - (,) TypePattern
TypePattern ::= Modifiersopt Type Identifier
----------------------------------------
Reduce       - ComponentPatternList
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - ()) TypePattern
TypePattern ::= Modifiersopt Type Identifier
             - ComponentPatternList
ComponentPatternList ::= ComponentPatternList COMMA...
----------------------------------------
Reduce       - ComponentPatternListopt
----------------------------------------
Shift/Reduce - (BeginCaseExpr) PushRPAREN
PushRPAREN ::= RPAREN
             - RecordPattern
RecordPattern ::= Modifiersopt ReferenceType PushLPAREN
             - CaseLabelElementPattern
----------------------------------------
Reduce       - CaseLabelElement
CaseLabelElement ::= CaseLabelElementPattern
----------------------------------------
Reduce       - SwitchLabelCaseLhs
SwitchLabelCaseLhs ::= case CaseLabelElements
----------------------------------------
Shift        - (->)
Shift/Reduce - ({) SwitchLabelExpr
SwitchLabelExpr ::= SwitchLabelCaseLhs BeginCaseExpr...
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (Identifier)
Shift        - (})
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
Shift        - (})
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
Shift        - (Unexpected End Of File)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        -
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (RestrictedIdentifierrecord)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (() RecordHeaderName1
RecordHeaderName1 ::= Modifiersopt...
----------------------------------------
Reduce       - RecordHeaderName
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - (,)
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - RecordComponent
RecordComponent ::= Modifiersopt Type...
----------------------------------------
Reduce       - RecordComponents
----------------------------------------
Shift        - (int)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - RecordComponent
RecordComponent ::= Modifiersopt Type...
             - RecordComponents
RecordComponents ::= RecordComponents COMMA...
----------------------------------------
Reduce       - RecordComponentsopt
----------------------------------------
Shift/Reduce - (implements) RecordComponentHeaderRightParen
RecordComponentHeaderRightParen ::= RPAREN
             - RecordHeader
RecordHeader ::= LPAREN RecordComponentsopt...
----------------------------------------
Shift        - (Identifier)
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - InterfaceType
InterfaceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - ClassHeaderImplements
ClassHeaderImplements ::= implements InterfaceTypeList
             - RecordHeaderPart
RecordHeaderPart ::= RecordHeaderName RecordHeader...
----------------------------------------
Shift        - (})
Reduce       - RecordBodyDeclarationopt
RecordBodyDeclarationopt ::=
----------------------------------------
Shift/Reduce - (Unexpected End Of File) RecordBody
RecordBody ::= LBRACE RecordBodyDeclarationopt RBRACE
             - RecordDeclaration
RecordDeclaration ::= RecordHeaderPart RecordBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (Unexpected End Of File)
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (public)
Reduce       - EnterCompilationUnit
EnterCompilationUnit ::=
----------------------------------------
Shift        - (class)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - ({) ClassHeaderName1
ClassHeaderName1 ::= Modifiersopt class Identifier
----------------------------------------
Reduce       - ClassHeaderName
----------------------------------------
Reduce       - ClassHeaderExtendsopt
----------------------------------------
Reduce       - ClassHeaderImplementsopt
----------------------------------------
Reduce       - ClassHeaderPermittedSubclassesopt
             - ClassHeader
ClassHeader ::= ClassHeaderName ClassHeaderExtendsopt...
----------------------------------------
Shift        - (public)
Reduce       - NestedType
NestedType ::=
----------------------------------------
Shift        - (void)
Reduce       - Modifiersopt
Modifiersopt ::= Modifiers
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Shift/Reduce - (Identifier) PrimitiveType
----------------------------------------
Reduce       - Type
Type ::= PrimitiveType
----------------------------------------
Shift        - (()
Shift/Reduce - (Identifier) MethodHeaderName
MethodHeaderName ::= Modifiersopt Type Identifier LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (.)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (.) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - Name
----------------------------------------
Shift        - (Identifier)
Shift/Reduce - (Identifier) QualifiedName
QualifiedName ::= Name DOT SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift        - ())
Reduce       - Dimsopt
Dimsopt ::=
             - VariableDeclaratorId
             - VariableDeclaratorIdOrThis
VariableDeclaratorIdOrThis ::= VariableDeclaratorId
             - FormalParameter
FormalParameter ::= Modifiersopt Type...
----------------------------------------
Reduce       - FormalParameterListopt
----------------------------------------
Shift/Reduce - ({) MethodHeaderRightParen
MethodHeaderRightParen ::= RPAREN
----------------------------------------
Reduce       - Dimsopt
Dimsopt ::=
             - MethodHeaderExtendedDims
MethodHeaderExtendedDims ::= Dimsopt
----------------------------------------
Reduce       - MethodHeaderThrowsClauseopt
             - MethodHeader
MethodHeader ::= MethodHeaderName FormalParameterListopt
----------------------------------------
Reduce       - NestedMethod
NestedMethod ::=
----------------------------------------
Shift        - (})
Reduce       - BlockStatementsopt
BlockStatementsopt ::=
----------------------------------------
Shift/Reduce - (}) MethodBody
MethodBody ::= NestedMethod LBRACE BlockStatementsopt...
             - MethodDeclaration
MethodDeclaration ::= MethodHeader MethodBody
----------------------------------------
Reduce       - ClassBodyDeclarations
----------------------------------------
Reduce       - ClassBodyDeclarationsopt
ClassBodyDeclarationsopt ::= NestedType...
----------------------------------------
Shift/Reduce - (Unexpected End Of File) ClassBody
             - ClassDeclaration
ClassDeclaration ::= ClassHeader ClassBody
----------------------------------------
Reduce       - InternalCompilationUnit
InternalCompilationUnit ::=...
             - CompilationUnit
CompilationUnit ::= EnterCompilationUnit...
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
-- ENTER INSIDE PARSE METHOD --
- Start --------------------------------
Shift        - (switch)
Shift        - (()
Shift        - (Identifier)
Shift        - ())
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Reduce       - AssignmentExpression
             - Expression
Expression ::= AssignmentExpression
----------------------------------------
Shift        - ({)
Reduce       - PostExpressionInSwitchStatement
PostExpressionInSwitchStatement ::=
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (case)
Shift        - (BeginCaseElement)
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (()
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Shift/Reduce - (Identifier) PushLPAREN
PushLPAREN ::= LPAREN
----------------------------------------
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - (,) TypePattern
TypePattern ::= Modifiersopt Type Identifier
----------------------------------------
Reduce       - ComponentPatternList
----------------------------------------
Shift        - (Identifier)
Reduce       - Modifiersopt
Modifiersopt ::=
----------------------------------------
Shift        - (Identifier)
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - ClassOrInterface
ClassOrInterface ::= Name
----------------------------------------
Reduce       - ClassOrInterfaceType
             - ReferenceType
ReferenceType ::= ClassOrInterfaceType
----------------------------------------
Reduce       - Type
----------------------------------------
Shift/Reduce - ()) TypePattern
TypePattern ::= Modifiersopt Type Identifier
             - ComponentPatternList
ComponentPatternList ::= ComponentPatternList COMMA...
----------------------------------------
Reduce       - ComponentPatternListopt
----------------------------------------
Shift/Reduce - (BeginCaseExpr) PushRPAREN
PushRPAREN ::= RPAREN
             - RecordPattern
RecordPattern ::= Modifiersopt ReferenceType PushLPAREN
             - CaseLabelElementPattern
----------------------------------------
Reduce       - CaseLabelElement
CaseLabelElement ::= CaseLabelElementPattern
----------------------------------------
Reduce       - SwitchLabelCaseLhs
SwitchLabelCaseLhs ::= case CaseLabelElements
----------------------------------------
Shift        - (->)
Shift/Reduce - ({) SwitchLabelExpr
SwitchLabelExpr ::= SwitchLabelCaseLhs BeginCaseExpr...
----------------------------------------
Reduce       - OpenBlock
OpenBlock ::=
----------------------------------------
Shift        - (Identifier)
Shift        - (})
Reduce       - Name
Name ::= SimpleName
----------------------------------------
Reduce       - PostfixExpression
PostfixExpression ::= Name
----------------------------------------
Error        -
Shift/Reduce - (}) EmptyStatement
EmptyStatement ::= SEMICOLON
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - BlockStatementsopt
----------------------------------------
Shift/Reduce - (Unexpected End Of File) Block
Block ::= OpenBlock LBRACE BlockStatementsopt RBRACE
             - SwitchLabeledBlock
SwitchLabeledBlock ::= SwitchLabelExpr Block
             - SwitchLabeledRule
----------------------------------------
Shift/Reduce - (Unexpected End Of File) SwitchBlock
             - SwitchStatement
SwitchStatement ::= switch LPAREN Expression RPAREN...
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - BlockStatementsopt
             - Goal
----------------------------------------
- End ----------------------------------
```

Reply:
> I tried, then the test fails, here is the parser output

Thanks, so yes this is more complex than I thought. Luckily the comparison of parser traces led me to where things went wrong:

Even after consuming the entire case statement, we still need to signal that extended recovery is needed to construct the enclosing switch.

To achieve this I invented yet another element kind: `K_ENHANCED_SWITCH`. As long as that kind is on the `elementStack` `requireExtendedRecovery()` will answer `true`.

OT1H I believe this status achieved in [36c3cf7](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2326/commits/36c3cf739f9bd9f5da04b36419a7365e31d35d1b) is better than before, because previously `K_BETWEEN_CASE_AND_COLON`, which was used to detect the need for extended recovery, remained on the stack only accidentally (`K_SELECTOR` and friends prevented us from `popElement(K_RECORD_PATTERN)`, disturbing also future pops).

Still I'd like to get a word from @srikanth-sankaran  because indeed we are relying on the machinery around `requireExtendedRecovery()` and `fallBackToSpringForward()`, which was designed with lambdas in mind, not for enhanced switches.

If mixing lambda recovery and switch recovery should better be separated, another totally different route might be inspired from `buildMoreCompletionContext()` branch `case K_BETWEEN_CASE_AND_COLON`. Why do I think so? Until now, CompletionParser doesn't care for any `SwitchStatement`, since previously the switch did not produce new interesting context. All we needed was the inner Block. That branch inside `buildMoreCompletionContext()` demonstrates how more context can be hand-crafted when needed. Here it's done for completion of the case constant, which needs the switch context indeed. We could develop a similar branch for the case where the assist node is inside the block of a pattern case. *Then* we would hopefully not need extended recovery.

@srikanth-sankaran what does the master of `fallBackToSpringForward()` recommend? :)

Reply:
@stephan-herrmann I added another tests which is a statement switch, seems still we have some issues when the switch is not a expression.

Reply:
Here's what I see with a ':' terminated case:
* initial parse goes as expected
* when error is detected and parser resumes in recovery mode 2 strange things happen:
  * kind `K_SWITCH_LABEL` is pushed *twice* on the element stack
  * when seeing `(` we don't find the expected `K_RECORD_PATTERN` and so proceed to push `K_SELECTOR` etc.

You may want to investigate, why the recovery parse doesn't produce `K_RECORD_PATTERN`

Perhaps `CompletionParser.recoveryTokenCheck()` is the culprit, which in its branch for `TokenNamecase` goes directly to `K_SWITCH_LABEL` skipping the sequence of `K_BETWEEN_CASE_AND_COLON` , `K_RECORD_PATTERN`?

Reply:
Isn't that happens when trying to parse the code in Unverified Variable Resolution attempts?

In the main parser i see that the parser exits just after seeing the first Rbrace of the case block. The token becomes EOF while the scanner shows the current position is still at the rbrace of the case block

Reply:
> Isn't that happens when trying to parse the code in Unverified Variable Resolution attempts?

what is "Unverified Variable Resolution"?

Reply:
Sorry mixed up names, UnresolvedReferenceNameFinder which is used since the current scope ends up with unresolved local variables from the incomplete ast tree.

Reply:
> Sorry mixed up names, UnresolvedReferenceNameFinder which is used since the current scope ends up with unresolved local variables from the incomplete ast tree.

My point is: the AST should be more complete (and it is in the `->` case).

Reply:
Next finding when stopping at `CompletionParser.endParse()`: even in the `->` case at `endParse()` the assembled switch statement still sits on the astStack and has not yet been transferred into the method (where until then we see the individual parts of the switch).  It's only in `AssistParser.parseBlockStatements()` where the assembled switch is transferred into the method.

In the `:` case, however, we never create a `SwitchStatement`, relevant sections of parser debug log:

Works:
```
Error        - 
Shift/Reduce - (}) EmptyStatement
EmptyStatement ::= SEMICOLON
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - BlockStatementsopt
----------------------------------------
Shift/Reduce - (Unexpected End Of File) Block
Block ::= OpenBlock LBRACE BlockStatementsopt RBRACE
             - SwitchLabeledBlock
SwitchLabeledBlock ::= SwitchLabelExpr Block
             - SwitchLabeledRule
----------------------------------------
Shift/Reduce - (Unexpected End Of File) SwitchBlock
             - SwitchStatement
SwitchStatement ::= switch LPAREN Expression RPAREN...
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - BlockStatementsopt
             - Goal
----------------------------------------
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
```
Works not:
```
Error        - 
Shift/Reduce - (}) EmptyStatement
EmptyStatement ::= SEMICOLON
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - BlockStatementsopt
----------------------------------------
Shift/Reduce - (Unexpected End Of File) Block
Block ::= OpenBlock LBRACE BlockStatementsopt RBRACE
----------------------------------------
Reduce       - BlockStatements
BlockStatements ::= BlockStatement
----------------------------------------
Reduce       - TypeAnnotationsopt
TypeAnnotationsopt ::=
----------------------------------------
Error        - 
- End ----------------------------------
-- EXIT FROM PARSE METHOD --
```

I haven't yet found why the second log takes a quick route to "Error" while the other still shows useful activity. I don't think that our handling of `elementKindStack` is to blame, there are small differences between both cases, but those diffs don't seem to affect the outcome.

Reply:
I did some checks for a while this evening, What I have found so far is the Parser doesn't find any switch related elements, thats is not of the `consumeSwitch*` are called after `consumeSwitchLabelCaseLhs`. I wonder if it is a scanner issue, I saw that there are some special handling for ARROW and Patterns, like the usages of this method `disambiguateArrowWithCaseExpr`, I don't see they have support for COLON. But not sure since still I cannot wrap my head around this different level of parsing and manipulation in these 4 components involve in Completion like @stephan-herrmann , @mpalat and @srikanth-sankaran

Reply:
> I did some checks for a while this evening, What I have found so far is the Parser doesn't find any switch related elements, thats is not of the `consumeSwitch*` are called after `consumeSwitchLabelCaseLhs`.

Side-by-side debugging of versions with `:` and `->` showed that the difference is only in the actual parser state machine. A the some point during parsing, when using `:` `ntAction(..)` gives an error action, when using `->` that state machine continues until the entire switch is consumed. In the end this is a result of different grammar rules. It _might_ be that the synthetic token `BeginCaseExpr` has a finger in the pie. But that's not so easy to figure out.

Given that normal parsing of all this, including recovery, is already black magic, I felt it's time to try something different, that does not rely on normal `requireExtendedRecovery()` plus regular parsing. In particular I wouldn't want to add complexity to the grammar just to improve parsing of incomplete code. @srikanth-sankaran  fyi :)

See #2391 for a fresh new attempt.

Reply:
To clarify, @stephan-herrmann when you say
> In the end this is a result of different grammar rules

Do we have different rules for completion parser or is it that rules get in our way when we are parsing a incomplete code, where the structure is not correctly parsed ?

Reply:
The difference is between variants using `:` vs `->`.

While both variant can of course be correctly parsed in complete code, completion limits source range to end at the cursor position, so yes, the problem is about incomplete code.

We don't directly control, how the parser automaton behaves when it sees the artificial EOF in the middle of a structure. Apparently, one variant still continues to reduce what's already on stacks and combines it into a SwitchStatement, while the other variant stops more or less immediately.